### ISSUE TITLE
Extend ContainerEngine support for scripting languages like Python, NodeJS, Ruby on rails.

### ISSUE THREAD

**Issue by a0xnirudh** [Hackademic/hackademic#181]
As of now the ContainerEngine supports only PHP based challenges. We should extend this further to support challenges in Python, NodeJS, Ruby on rails, and if possible Java.

The extension should be made in such a way that ContainerEngine should work smoothly and should not take too much memory (this should be discussed and investigated further.)

**Comment by northdpole** [comment]
the engine doesn't support anything, the example challenge and the dockerfile support and install php, however we could provide a docker file for whatever we support or let the user provide docker related instructions in order to abstract this step, right?

**Comment by a0xnirudh** [comment]
Yes we use docker file to install things as we need. So for running python/node js, we can add simple  2 - 3 steps to docker file and its done but their is a problem. Once the thing is installed, we will also initiate all of them by adding them to init script. So some amount of ram will be simply wasted without running anything, like for example, for running a nodejs challenge, we don't need apache running but its difficult to configure it that way inside the container.

Also according to the docker image documentation regarding the `init script`, the installed programs should run in daemonised mode, i.e., the process should create children or new process by itself (apache will run like this if we call it with -X parameter like specified in the docker file right now). I am not familiar with NodeJS or ruby on rails so before adding these to docker file, we have to make sure its running just the way we want.

If we have solutions to above 2 problems, then this issue is solved. :)

**Comment by AmaanC** [comment]
I'd like to work on adding support for NodeJS (i.e. modifying the Dockerfile and creating a sample challenge). Assuming no one's already working on that?

**Comment by a0xnirudh** [comment]
@AmaanC I am already working on it. I just pushed the support to Ruby on Rails last day and I will push the support to Node.Js soon (Almost complete). Probably you can work on Java or other languages. 

Also make sure you discuss the idea on how  you are planning to embedded it first. For example, the main [Dockerfile](https://github.com/Hackademic/hackademic/blob/a0xnirudh/docker/ContainerManager/Dockerfile) is no the place where you add these. I have added Ruby on Rails to it because its a POC for my GSoC proposal. Docker build should happen dynamically, probably based on the same base image and not the one where PHP/Apache/MySQL is installed ! So challenge authors has to add a custom Dockerfile along with their challenges (There are some more challenges to this method which we are trying to solve now).

**Comment by AmaanC** [comment]
Ah, that sounds fair. I'll look into some of the other issues instead.
